client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientFactoryImpl.java

```diff
@@ -74,4 +78,12 @@ protected void onDisconnect() {
     private void removeDeviceConnectionClient(final String tenantId) {
         deviceConnectionClientFactory.removeClient(DeviceConnectionClientImpl.getTargetAddress(tenantId));
     }
+
+    private void handleTenantTimeout(final Message<String> msg) {
```

Comment:
There are 4 places with a similar `handleTenantTimeout` method now. I had a look here and could imagine getting rid of the code duplication with a method like:
````
    private static <T extends RequestResponseClient> void closeAndRemoveClient(
            final CachingClientFactory<T> cachingClientFactory, Supplier<String> clientKeySupplier) {
        final String key = clientKeySupplier.get();
        final T client = cachingClientFactory.getClient(key);
        if (client != null) {
            client.close(v -> cachingClientFactory.removeClient(key));
        }
    }
````
invoking it like this:
````
        connection.getVertx().eventBus().consumer(Constants.EVENT_BUS_ADDRESS_TENANT_TIMED_OUT,
                msg -> {
                    final String tenantId = msg.body().toString();
                    closeAndRemoveClient(deviceConnectionClientFactory,
                            () -> DeviceConnectionClientImpl.getTargetAddress(tenantId));
                });
````
But this would still look like an intermediate step to me, as I could imagine having a `CachingClientFactory.closeAndRemoveClient(Supplier<String> clientKeySupplier)` in the end. For this, the `CachingClientFactory<T>` class has to be changed so that the `T` extends an interface with a `close` method though. So refactoring the `handleTenantTimeout` methods could be postponed until after the `CachingClientFactory` has been changed. WDYT?

Reply:
Yes, I would welcome a refactoring that minimizes code duplication and agree that this should be done later on and not within this PR.